### PR TITLE
ENGINES: Streamline auto-save write/delete protection

### PR DESCRIPTION
**Pull request by orgads**
Some engines call setAutosave and some don't. isAutosave is used to
determine if a saved game is an autosave, but in fact, on most cases it
just falls back to comparing the name to "Autosave".

This is wrong for several reasons:
* Older versions of ScummVM used Autosave 0.
* The name "Autosave" is translated, so if you change the language, it
  won't be detected.

Instead of relying on the name, use the well-known getAutosaveSlot() from
Engine/MetaEngine.

Fixes #12735.

### CODE REVIEW

**Comment by lephilousophe** [comment]
getAutosaveSlot() tells where autosaves are supposed to be saved.
If a user has saved on the autosave slot, it won't be an autosave and I am not sure what will happen with your changes

**Comment by orgads** [comment]
Most (all?) engines write-protect and delete-protect this entry, so the user cannot save on it.

**Comment by orgads** [comment]
Rebased to trigger CI.

**Comment by orgads** [comment]
Ok, I discovered the global `g_engine`, and simplified this patch a lot. Please review.

**Comment by lephilousophe** [comment]
> Most (all?) engines write-protect and delete-protect this entry, so the user cannot save on it.

This is not the case on at least Versailles. I always thought that autosaving support were optional and never did anything to implement it.
This is a quite a huge change so I think some stance should be taken on this.

**Comment by orgads** [comment]
> This is a quite a huge change so I think some stance should be taken on this.

It's much smaller now 🙂

Did you see the latest revision?

File: engines/savestate.cpp
Comment (lephilousophe) [inline engines/savestate.cpp:49]: This must take autosave_period configuration into account like it's done in MetaEngine::listSaves

Comment (orgads) [reply engines/savestate.cpp:49]: I thought about it, but I think it's better not to. Consider the following use-case:

* Autosave is disabled - slot 0 is a regular slot.
* The user happily saves on this slot.
* After several weeks, the user discovers the autosave feature and enabled it.
* 5 minutes into the game and Poof - the saved game is gone!

Comment (sev-) [reply engines/savestate.cpp:49]: In this case it would be best to warn the user when they enable this option, something like: "The autosave uses slot XX, it will be lost. Proceed?"

Comment (orgads) [reply engines/savestate.cpp:49]: Good point. So when enabling this option, I'll iterate over all the engines and check if they have a saved game which is not named "Autosave" on the autosave slot, and warn on this case?

Comment (sev-) [reply engines/savestate.cpp:49]: Exactly, something like this.

**Review by sev- — APPROVED** [review]
This is a good and simple change, and we could still squeeze it into the release because it is easy to check and it is obvious when it is broken or not.

So, please, implement more sanity check when switching the option and we merge it.

File: engines/savestate.cpp
Comment (orgads) [reply engines/savestate.cpp:49]: I have another idea, which I'm working on.

When autosave is triggered, I check if this slot is occupied by a non-autosave entry. If it is, I prompt the user to choose between:
* Move the currently saved game to a free slot
* Overwrite
* Disable autosave (for this session)

A similar approach can also be useful to prevent accidental overwrite of autosave. For example, if the existing autosave was saved after 3 hours of gameplay, then the user started a new game and autosave was triggered, we can warn before overwriting the existing one.

What do you say?

Comment (orgads) [reply engines/savestate.cpp]: Done

File: engines/savestate.cpp
Comment (orgads) [reply engines/savestate.cpp]: Done

**Comment by orgads** [comment]
Factored out the prompt as a function, and changed the check to be contains("Autosave") instead of equals, because previous versions of ScummVM saved it as "Autosave 0".

**Comment by orgads** [comment]
@sev- I have another pending commit for warning on options change. https://github.com/orgads/scummvm/tree/autosave-improved

Waiting for final approval of this PR before opening a new PR for that.

**Comment by macca8** [comment]
Activating a warning message if the slot is deemed to be occupied by a non-autosave saved game at the time the autosave is triggered is a great inclusion, however, allowing the player to disable autosaving as an option seems odd. 
Isn't the objective here to encourage the player to take some action, when necessary, to allow the autosave to proceed?

At worst, I'd recommend that 'Cancel autosave' should only cancel the current autosave (the message could be simplified to "or cancel this autosave."). By default, failed autosaves are repeated at 5 minute intervals (which covers all other 'saveFlag = false' instances). It makes sense for this option to follow the same pattern.

**Comment by orgads** [comment]
Thanks for the feedback.

Pressing Escape cancels the current save, but will prompt again next time. I know this is not so discoverable :)

Do you think it's worth adding another button for this action? I don't want to clutter the dialog too much.

**Comment by sev-** [comment]
Thank you, merging.

@macca8 temporary disabling autosave for a session when there is a problem seems like the most logical approach to me, instead of getting nagged every 5 minutes.

**Comment by orgads** [comment]
Great! Follow-up PR: #3304

**Pull request merged**
